pkg/election/leadership.go

```diff
@@ -182,26 +188,69 @@ func (ls *Leadership) Watch(serverCtx context.Context, revision int64) {
 	if ls == nil {
 		return
 	}
+
+	interval := detectHealthyInterval
+	timeout := watchLoopUnhealthyTimeout
+	failpoint.Inject("fastTick", func() {
+		timeout = 5 * time.Second
+		interval = 1 * time.Second
+	})
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+	lastHealthyTime := time.Now()
+
 	watcher := clientv3.NewWatcher(ls.client)
 	defer watcher.Close()
-	ctx, cancel := context.WithCancel(serverCtx)
-	defer cancel()
-	// The revision is the revision of last modification on this key.
-	// If the revision is compacted, will meet required revision has been compacted error.
-	// In this case, use the compact revision to re-watch the key.
 	for {
 		failpoint.Inject("delayWatcher", nil)
-		rch := watcher.Watch(ctx, ls.leaderKey, clientv3.WithRev(revision))
-		for wresp := range rch {
+		// In order to prevent a watch stream being stuck in a partitioned node,
+		// make sure to wrap context with "WithRequireLeader".
+		watchChanCtx, watchChanCancel := context.WithCancel(clientv3.WithRequireLeader(serverCtx))
+		defer watchChanCancel()
```

Comment:
watchChanCancel() is at places which needs to be handled very carefully and is error prone. We can make it much clearer and easier in the following way (BTW, cancel() can be called multiple times):
```suggestion
func GetLeader(c *clientv3.Client, leaderPath string) (*pdpb.Member, int64, error) {
    var watchChanCancel context.Cancel
    defer func() {
        if watchChanCancel != nil {
	    watchChanCancel()
	}
    } ()
    
    for {
        ...
        if watchChanCancel != nil {
            watchChanCancel()
         }
         watchChanCtx, watchChanCancel := context.WithCancel(clientv3.WithRequireLeader(serverCtx))
         ...
```

Reply:
Agree. Here we might have too many defer.

Reply:
ok, I have added it in https://github.com/tikv/pd/blob/5b3378404f35b1731845b1ac1253181d414037e4/pkg/election/leadership.go#L204